### DISCUSSION TITLE
Special characters in page paths

### DISCUSSION THREAD

**Discussion by Leksat** [gatsbyjs/gatsby#36345] [Help]
As far as I can see, Gatsby expects page paths to be _not_ encoded. Otherwise it produces double-encoded URLs.

If the paths are not encoded, Gatsby handles them well.
```ts
createPage({ path: '/page/🎉' })
```
https://github.com/Leksat/my-gatsby-site/commit/78597a39add9fe50eb368dc56f9fb79f472a00ef

<img width="983" alt="Screen Shot 2022-08-09 at 16 17 31" src="https://user-images.githubusercontent.com/989015/183658466-ae06aac4-eff4-4b3d-b463-a206b8a11991.png">

## Percent sign

If I try to create a page with a path containing `%` sign, the build will fail with `URIError: URI malformed`.

https://github.com/Leksat/my-gatsby-site/commit/5e54e7cf6aa2e4fac914d5c6a6b4c00d25af5e33

<img width="1286" alt="Screen Shot 2022-08-09 at 16 36 53" src="https://user-images.githubusercontent.com/989015/183664791-25359780-175a-466b-bf44-0dd51fb7952d.png">

(This is because `decodeURIComponent` is applied to paths in `@gatsbyjs/reach-router`:
https://github.com/gatsbyjs/reach-router/blob/ccbdf3e10d6501093abd4f15e80f6261e5c7b036/src/lib/utils.js#L68)

If I try to encode the `%` sign as `%25`, the result will be buggy.

```ts
createPage({ path: '/page-with-%25-sign' })
```

http://localhost:9000/page-with-%25-sign will be 404

http://localhost:9000/page-with-%2525-sign will work, but the URL in the browser address bar will be rewritten to http://localhost:9000/page-with-%25-sign (why?), so it will be 404 once you reload the page

Is this a bug in `@gatsbyjs/reach-router`? Or am I doing something wrong?

**UPD:** submitted https://github.com/gatsbyjs/gatsby/issues/36350 as it really looks like a bug.

## Other special characters

I tried to go with a crazy path.
```ts
createPage({ path: '/page/!#$^&*(){}[]:"|;\'\\<>?,./`~-=' })
```
https://github.com/Leksat/my-gatsby-site/commit/4682a880eec44aeb58dae93e3cccd22b818e9a53

The build went fine, but `gatsby serve` could not find the page, even if the dirs and the file were in place.

<img width="1145" alt="Screen Shot 2022-08-09 at 16 22 14" src="https://user-images.githubusercontent.com/989015/183661052-af82a3b7-a077-40ad-8be7-eb8d36da00cb.png">

Why Gatsby couldn't serve the URL? Is there a list of characters which I should avoid in paths?

**UPD:** I did some experiments. Found that the following characters are not safe for Gatsby paths: `#`, `?`, `\`.

**UPD2**:
- `#` and `?` must be encoded in visited URL for obvious reason - they are control URL characters splitting hash and query from the path. If I encode them, the URL works (almost).
- `\` must be encoded as well, not sure why. But I guess it's better to not use it in general to avoid issues on Windows.
- `"` breaks JS due to a bug, I opened an issue: https://github.com/gatsbyjs/gatsby/issues/36348

---

In general: How to properly handle special chars in Gatsby page paths?

(Thanks for reading! Sorry for a lot of words 😅)

**Comment by tyhopp** [comment]
There was a recent docs PR on a similar topic (https://github.com/gatsbyjs/gatsby/pull/36325). If you're interested in contributing any clarifying details of what's permissible in paths that could also be useful!

**Comment by Leksat** [comment]
This is a rabbit hole 😅 

I submitted a couple of bugs, but there are still some edge cases. For example, `/my/../page` path:
- Gatsby will create `page/index.html` file
- if I try to visit `/my/../page` I'll be redirected to `/page`
- but there will be `Hydration failed` error (the page will work without JS enabled though)

I'm sure I can find more, but I think I should stop here.

The general recommendation for Gatsby page paths would be:
- Avoid using `"` until https://github.com/gatsbyjs/gatsby/issues/36348 is solved
- Avoid using `%` until https://github.com/gatsbyjs/gatsby/issues/36350 is solved
- Avoid using `#`, `?` and `\` just in case (to avoid URL encoding/decoding issues)
- Avoid using dots-only path segments
